### PR TITLE
Implement IEquatable for DateInterval.

### LINKED ISSUE #380: Make DateInterval IEquatable
This is a feature request to make DateInterval implement IEquatable

### PR DESCRIPTION
**Pull request by malcolmr**
Also implement `GetHashCode()` and a basic `ToString()`.

Fixes #380.

### CODE REVIEW

File: src/NodaTime.Test/DateIntervalTest.cs [hunk 1]
```diff
@@ -60,6 +60,117 @@ public void Construction_Properties()
...
+        [Test]
+        public void Equals_DifferentStart()
+        {
+            LocalDate start1 = new LocalDate(2000, 1, 1);
+            LocalDate start2 = new LocalDate(2000, 1, 2);
+            LocalDate end = new LocalDate(2001, 6, 19);
+            var interval1 = new DateInterval(start1, end, false);
+            var interval2 = new DateInterval(start2, end, false);
+            Assert.AreNotEqual(interval1, interval2);
+            Assert.AreNotEqual(interval1.GetHashCode(), interval2.GetHashCode());
```
Comment (malcolmr) [inline src/NodaTime.Test/DateIntervalTest.cs]: This doesn't seem valid, by the way: there's generally no guarantee that two hash codes will be non-equal in the event that the instances are non-equal. However, we seem to do this elsewhere as well, so perhaps I'm missing something?

File: src/NodaTime/DateInterval.cs [hunk 2]
```diff
@@ -87,6 +92,73 @@ public DateInterval(LocalDate start, LocalDate end)
...
+        /// <param name="lhs">The first value to compare</param>
+        /// <param name="rhs">The second value to compare</param>
+        /// <returns>True if the two date intervals have the same properties; false otherwise.</returns>
+        public static bool operator ==(DateInterval lhs, DateInterval rhs)
+        {
+            if (ReferenceEquals(lhs, rhs))
+            {
+                return true;
+            }
+            if (ReferenceEquals(lhs, null) || ReferenceEquals(rhs, null))
```
Comment (malcolmr) [inline src/NodaTime/DateInterval.cs]: I assume that this is correct, but I note that we don't implement `operator==` for any other classes, so I don't have anything to check against.

File: src/NodaTime.Test/DateIntervalTest.cs (hunk 1 above)
Comment (jskeet) [inline src/NodaTime.Test/DateIntervalTest.cs]: There's no guarantee, but if the test fails, the chances of it being due to a bug are extremely high.

File: src/NodaTime/DateInterval.cs (hunk 2 above)
Comment (jskeet) [inline src/NodaTime/DateInterval.cs]: Yup, that's fine. An alternative approach is to make `==` call `Equals`, leading to just one nullity check. But this feels nicer.

**Comment by jskeet** [comment]
Right, will merge this in and then address my own comments :)

**Pull request merged**